libcst/codemod/visitors/_add_imports.py

```diff
@@ -8,11 +8,47 @@
 
 import libcst
 from libcst import matchers as m, parse_statement
+from libcst._nodes.statement import Import, ImportFrom, SimpleStatementLine
 from libcst.codemod._context import CodemodContext
 from libcst.codemod._visitor import ContextAwareTransformer
 from libcst.codemod.visitors._gather_imports import GatherImportsVisitor
 from libcst.codemod.visitors._imports import ImportItem
 from libcst.helpers import get_absolute_module_from_package_for_import
+from libcst.helpers.common import ensure_type
+
+
+class _GatherTopImportsBeforeStatements(GatherImportsVisitor):
+    """
+    Works similarly to GatherImportsVisitor, but only considers imports
+    declared before any other statements of the module with the exception
+    of docstrings and __strict__ flag.
+    """
+
+    def leave_Module(self, original_node: libcst.Module) -> None:
```

Comment:
This isn't a great use of inheritance, `_GatherTopImportsBeforeStatements` doesn't really specialize the behavior of `GatherImportsVisitor`, it completely changes it.

A better way to model this would be to extract the two `_handle_...` methods into a mixin class and let both visitors inherit from that.

Then `_GatherTopImportsBeforeStatements` won't need to override `visit_ImportFrom` and `visit_Import` with dummy implementations (and potentially others in the future). Instead you could just implement `visit_Module` and return `False` at the end. Bonus points for making it significantly faster for large documents.